csrc/custom_all_reduce.cu

```diff
@@ -154,16 +154,20 @@ void register_graph_buffers(fptr_t _fa, const std::vector<std::string>& handles,
 
 #ifdef USE_ROCM
 
-void free_meta_buffer(void* buffer) { hipFree(buffer); }
+void free_meta_buffer(void* buffer) { CUDACHECK(cudaFree(buffer)); }
 
-std::vector<uint8_t> get_meta_buffer_ipc_handle(torch::Tensor inp) {
-  std::vector<uint8_t> data_handle(sizeof(cudaIpcMemHandle_t), 0);
-  CUDACHECK(cudaIpcGetMemHandle((cudaIpcMemHandle_t*)data_handle.data(),
+torch::Tensor get_meta_buffer_ipc_handle(torch::Tensor& inp) {
```

Comment:
Could you elaborate on this? Why the change to tensor and why its python counterpart still marks the return type as List[str]

Reply:
I'm following the discussion in https://github.com/vllm-project/vllm/pull/5047 where returning a `std::vector<uint8_t>` is not supported in some circumstances, seemingly because Torch Extension doesn't have full support for `uint8_t`. Specific comment: https://github.com/vllm-project/vllm/pull/5047#issuecomment-2140709396

I'm making this change to follow the convention of `get_graph_buffer_ipc_meta`, which originally returned a C++ `pair<vector<uint8_t>, vector<int64_t>>` -> Python `Tuple[List[str], List[int]]` (seemingly duck-typing a `uint8_t` with a char) and now returns a `pair<Tensor (of type uint8), vector<int64_t>>` but preserving the same Python type hint.

I'm not convinced that this is strictly speaking correct, but it does follow upstream convention and works at present. As per the comments in the original PR, if this is wrong, we are likely to see evidence only when `torch.compile` is used more liberally and we start seeing problems here.

**Extra context:**
Some aspects of the Torch Extension system are not easy to understand and are not well-documented. I'm not convinced that PR gets everything right: for instance, it often does something like the following
```
  custom_ar.def("get_graph_buffer_ipc_meta", &get_graph_buffer_ipc_meta);
  custom_ar.impl("get_graph_buffer_ipc_meta", torch::kCPU, &get_graph_buffer_ipc_meta);
```
But it doesn't seem the second line is necessary: the original `.def` **with the optional function pointer supplied** already (implicitly) registers this method for **all** dispatch keys (including `torch::kCUDA`, `torch::kCPU`, etc.). The subsequent `.impl` registers this method with the CPU dispatch key and hence appears to be redundant.

In general we would expect `.impl` to be redundant if we are not intending on having different implementations of a method depending on the device (e.g. different CPU and GPU methods for computing sine of a Tensor). Put another way, we should never need to use `.impl` if we only ever have/expect a single implementation.

For CAR, which is currently NV/AMD GPU-only, the developer knows at compile-time whether a Tensor is on CPU or GPU, and accordingly we only have one implementation of each method. So we don't need the `.impl`-s. This might change in the future if say someone generalizes CAR to also work on CPU-only vLLM, and hence they need to provide implementations of these methods for the cases where some of the all reduce buffers are on CPU (instead of being on GPU).

Reply:
Some documentation for details about custom operator dispatch: https://pytorch.org/cppdocs/api/file_torch_library.h.html#file-torch-library-h
https://pytorch.org/cppdocs/library.html#classtorch_1_1_library_1a59330802ab9deaff247d32a22092bdfb

When dispatching happens:
> When you pass operators with tensors of your custom backend [editor note: corresponding to a particular dispatch key], your overridden implementations will be called instead of the standard implementations

Can't say I understand the dispatch system well, but seems to mean something like: if a `torch.Tensor` is passed as an argument, the method it's dispatched to is one corresponding to the dispatch key of the Tensor.

Using `.def` with a function pointer, schema optional:
> Define an operator for a schema and then register an implementation for it.

This is typically what you would use if you aren’t planning on making use of the dispatcher to structure your operator implementation. It’s roughly equivalent to calling [def()](https://pytorch.org/cppdocs/library.html#classtorch_1_1_library_1a983e5fef5b234ad18cf389ed820e33c0) and then [impl()](https://pytorch.org/cppdocs/library.html#classtorch_1_1_library_1a59330802ab9deaff247d32a22092bdfb), but if you omit the schema of the operator, we will infer it from the type of your C++ function. All template arguments are inferred.

Using `.def` without a function pointer:
> Declare an operator with a schema, but don’t provide any implementations for it.

You’re expected to then provide implementations using the [impl()](https://pytorch.org/cppdocs/library.html#classtorch_1_1_library_1a59330802ab9deaff247d32a22092bdfb) method. All template arguments are inferred.


Using `.impl`:

> Register an implementation for an operator.

You may register multiple implementations for a single operator at different dispatch keys (see [torch::dispatch()](https://pytorch.org/cppdocs/library.html#group__torch-dispatch-overloads_1ga91120ca3079a024775f01ea37e7fcbd8)). Implementations must have a corresponding declaration (from [def()](https://pytorch.org/cppdocs/library.html#classtorch_1_1_library_1a983e5fef5b234ad18cf389ed820e33c0)), otherwise they are invalid. **If you plan to register multiple implementations, DO NOT provide a function implementation when you [def()]** (https://pytorch.org/cppdocs/library.html#classtorch_1_1_library_1a983e5fef5b234ad18cf389ed820e33c0) the operator.